<a href="https://colab.research.google.com/github/YefridC09/ST-554-Project1-Template/blob/main/Task2/st554_project1_task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Author: Eleanor Warren
###Date: 2/17/2026
###Purpose: First Draft

Time series data for air quality measurements in Italy.

- True measurements of air pollutants
- Sensor measurements of pollutants
- Weather-related measurements

This notebook should include:
- Introduction/narrative/section headers/charts and graphs

In [3]:
!pip install ucimlrepo

## modify column names or do other manipulations

In [48]:
from ucimlrepo import fetch_ucirepo
import numpy as np
import pandas as pd

# Keep air_quality as our repo dictionary
air_quality = fetch_ucirepo(id=360)
type(air_quality)

ucimlrepo.dotdict.dotdict

In [36]:
print(air_quality.metadata)

# This includes metadata including the unique UCI dataset identifier,
# name and data description, repo link, raw data file link, abstract,
# it seems that they use the term "features" to describe variables.
# This metadata says there are no missing values but the paper description says it does?

{'uci_id': 360, 'name': 'Air Quality', 'repository_url': 'https://archive.ics.uci.edu/dataset/360/air+quality', 'data_url': 'https://archive.ics.uci.edu/static/public/360/data.csv', 'abstract': 'Contains the responses of a gas multisensor device deployed on the field in an Italian city. Hourly responses averages are recorded along with gas concentrations references from a certified analyzer. ', 'area': 'Computer Science', 'tasks': ['Regression'], 'characteristics': ['Multivariate', 'Time-Series'], 'num_instances': 9358, 'num_features': 15, 'feature_types': ['Real'], 'demographics': [], 'target_col': None, 'index_col': None, 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_of_dataset_creation': 2008, 'last_updated': 'Sun Mar 10 2024', 'dataset_doi': '10.24432/C59K5F', 'creators': ['Saverio Vito'], 'intro_paper': {'ID': 420, 'type': 'NATIVE', 'title': 'On field calibration of an electronic nose for benzene estimation in an urban pollution monitoring scenario', 'authors': 

In [45]:
# This is great because it contains the units
print(air_quality.variables)

             name     role         type demographic  \
0            Date  Feature         Date        None   
1            Time  Feature  Categorical        None   
2          CO(GT)  Feature      Integer        None   
3     PT08.S1(CO)  Feature  Categorical        None   
4        NMHC(GT)  Feature      Integer        None   
5        C6H6(GT)  Feature   Continuous        None   
6   PT08.S2(NMHC)  Feature  Categorical        None   
7         NOx(GT)  Feature      Integer        None   
8    PT08.S3(NOx)  Feature  Categorical        None   
9         NO2(GT)  Feature      Integer        None   
10   PT08.S4(NO2)  Feature  Categorical        None   
11    PT08.S5(O3)  Feature  Categorical        None   
12              T  Feature   Continuous        None   
13             RH  Feature   Continuous        None   
14             AH  Feature   Continuous        None   

                                          description       units  \
0                                                N

In [39]:
air_data = air_quality.data.features

# What type is our data set?
type(air_data) #Pandas Dataframe - this lets us know how to treat it.

pandas.core.frame.DataFrame

In [40]:
# Recode missing values from -200 to NaN
air_data = air_data.replace(to_replace = -200, value = np.nan)

# Check out first few values
air_data.head()

,Date,Time,CO(GT),PT08.S1(CO),NMHC(GT),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),PT08.S4(NO2),PT08.S5(O3),T,RH,AH
0,3/10/2004,18:00:00,2.6,1360.0,150.0,11.9,1046.0,166.0,1056.0,113.0,1692.0,1268.0,13.6,48.9,0.7578
1,3/10/2004,19:00:00,2.0,1292.0,112.0,9.4,955.0,103.0,1174.0,92.0,1559.0,972.0,13.3,47.7,0.7255
2,3/10/2004,20:00:00,2.2,1402.0,88.0,9.0,939.0,131.0,1140.0,114.0,1555.0,1074.0,11.9,54.0,0.7502
3,3/10/2004,21:00:00,2.2,1376.0,80.0,9.2,948.0,172.0,1092.0,122.0,1584.0,1203.0,11.0,60.0,0.7867
4,3/10/2004,22:00:00,1.6,1272.0,51.0,6.5,836.0,131.0,1205.0,116.0,1490.0,1110.0,11.2,59.6,0.7888


In [41]:
# Get the list of variables in the data set
air_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9357 entries, 0 to 9356
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           9357 non-null   object 
 1   Time           9357 non-null   object 
 2   CO(GT)         7674 non-null   float64
 3   PT08.S1(CO)    8991 non-null   float64
 4   NMHC(GT)       914 non-null    float64
 5   C6H6(GT)       8991 non-null   float64
 6   PT08.S2(NMHC)  8991 non-null   float64
 7   NOx(GT)        7718 non-null   float64
 8   PT08.S3(NOx)   8991 non-null   float64
 9   NO2(GT)        7715 non-null   float64
 10  PT08.S4(NO2)   8991 non-null   float64
 11  PT08.S5(O3)    8991 non-null   float64
 12  T              8991 non-null   float64
 13  RH             8991 non-null   float64
 14  AH             8991 non-null   float64
dtypes: float64(13), object(2)
memory usage: 1.1+ MB


In [65]:
# Get variable names in a more compatible format for python by removing certain symbols

# first rename the "TRUE" vars
air_data = air_data.rename(columns={"CO(GT)" : "CO", "NMHC(GT)" : "NMHC", \
                                    "C6H6(GT)" : "C6H6", "NOx(GT)" : "NOx", \
                                    "NO2(GT)" : "NO2"})
# rename the sensor vars
air_data = air_data.rename(columns={"PT08.S1(CO)" : "CO_S1", "PT08.S2(NMHC)" : "NMHC_S2", \
                                    "PT08.S3(NOx)" : "NOx_S3", "PT08.S4(NO2)" : "NO2_S4", \
                                    "PT08.S5(O3)" : "O3_S5"})
air_data.head()


,Date,Time,CO,CO_S1,NMHC,C6H6,NMHC_S2,NOx,NOx_S3,NO2,NO2_S4,O3_S5,T,RH,AH
0,3/10/2004,18:00:00,2.6,1360.0,150.0,11.9,1046.0,166.0,1056.0,113.0,1692.0,1268.0,13.6,48.9,0.7578
1,3/10/2004,19:00:00,2.0,1292.0,112.0,9.4,955.0,103.0,1174.0,92.0,1559.0,972.0,13.3,47.7,0.7255
2,3/10/2004,20:00:00,2.2,1402.0,88.0,9.0,939.0,131.0,1140.0,114.0,1555.0,1074.0,11.9,54.0,0.7502
3,3/10/2004,21:00:00,2.2,1376.0,80.0,9.2,948.0,172.0,1092.0,122.0,1584.0,1203.0,11.0,60.0,0.7867
4,3/10/2004,22:00:00,1.6,1272.0,51.0,6.5,836.0,131.0,1205.0,116.0,1490.0,1110.0,11.2,59.6,0.7888


In [66]:
# Quick overview analysis of all the numeric variables (everything but date and time)
air_data.describe()

,CO,CO_S1,NMHC,C6H6,NMHC_S2,NOx,NOx_S3,NO2,NO2_S4,O3_S5,T,RH,AH
count,7674.000000,8991.000000,914.000000,8991.000000,8991.000000,7718.000000,8991.000000,7715.000000,8991.000000,8991.000000,8991.000000,8991.000000,8991.000000
mean,2.152750,1099.833166,218.811816,10.083105,939.153376,246.896735,835.493605,113.091251,1456.264598,1022.906128,18.317829,49.234201,1.025530
std,1.453252,217.080037,204.459921,7.449820,266.831429,212.979168,256.817320,48.370108,346.206794,398.484288,8.832116,17.316892,0.403813
min,0.100000,647.000000,7.000000,0.100000,383.000000,2.000000,322.000000,2.000000,551.000000,221.000000,-1.900000,9.200000,0.184700
25%,1.100000,937.000000,67.000000,4.400000,734.500000,98.000000,658.000000,78.000000,1227.000000,731.500000,11.800000,35.800000,0.736800
50%,1.800000,1063.000000,150.000000,8.200000,909.000000,180.000000,806.000000,109.000000,1463.000000,963.000000,17.800000,49.600000,0.995400
75%,2.900000,1231.000000,297.000000,14.000000,1116.000000,326.000000,969.500000,142.000000,1674.000000,1273.500000,24.400000,62.500000,1.313700
max,11.900000,2040.000000,1189.000000,63.700000,2214.000000,1479.000000,2683.000000,340.000000,2775.000000,2523.000000,44.600000,88.700000,2.231000


In [67]:
# Get a sense of how many missing values are in this dataset
air_data.isna().sum()

,0
Date,0
Time,0
CO,1683
CO_S1,366
NMHC,8443
C6H6,366
NMHC_S2,366
NOx,1639
NOx_S3,366
NO2,1642


Task 2:

Understand the relationship that sensor and weather variables have with C6H6 (benzene).

I can ignore the "True" variables. I can also include additional relationships of interest in the variables.

Basic requirements:

- Numeric summaries of C6H6 (at different levels/combos of other variables)
- Investigate correlations
- Plots of C6H6
- Relationships over time and ignoring time.


## Reading the Paper and Taking Notes:

- Dataset contains 9358 instances of hourly averaged responses from an array of 5 metal oxide chemical sensors.
- Device was located in a significantly polluted area/road level/Italian city
- Timeline: March 2004 to Feb 2005 (1 year)
- "Ground Truth" hourly averaged conentrations of chems.
- Has missing values? Yes, represented by -200.
- This data was specifically used in an intro paper that wants to predict benzene concentration with the goal of monitoring urban envrionment pollution.
- Benzene is of interest because it's known to induce cancers.

- It seems like the goal of the paper is to show that you can use gas multi-sensors (electronic noses) to help gather more pollutant data.

- If certain pollutants are correlated, then they could estimate levels of pollutants that they aren't specifically measuring based on levels of a correlated pollutant.

- "We present the results of the evaluation of an on-field, multi-variate calibration of a multi-sensor device, based on seven solid-state sensors..." The work was targeted at the estimation of CO, NOx, NO2, and Benzene concentrations.

- The performance of their multi-sensor device is characterized by the estimation error with respect to conventional station output, investigating the influence of training length (does more time make it a better measure? Is there a period of time that provides optimal measures?)


Question 1: How is this data stored?

This data is stored as a pandas dataframe and access through a repo that contains a dictionary with values for data, metadata, and variable info.

There are 15 variables in this dataset.
